README.md
# Alan Italian

    Alan Standard Library: 2.1
    Alan Development Kit:  3.0beta6 build 1878

- https://github.com/tajmone/Alan3-Italian

|                                                 |
|-------------------------------------------------|
| **[CLICK HERE FOR ENGLISH README!][README_EN]** |

Questo progetto ha due scopi:

- Fornire supporto per la lingua italiana in Alan IF 3.
- Tradurre in italiano la _[Alan Standard Library]_ v2.1 di [Anssi Räisänen]. 

Alan Italian diventerà il sistema di autoraggio di AT italiane più facile da usare per i non programmatori, con una ricca libreria standard di partenza con oltre 170 verbi e una vasta gamma di classi predefinite e funzionalità per creare personaggi, fonti di luce, dispositivi, vestiario, porte e finestre, serrature, e molto altro ancora.

[![Donazione PayPal][donazione img]][donazione link]

Progetto a cura di [Tristano Ajmone].

[Licenza][License]: Artistic License 2.0

Data inizio progetto: 2018/04/13.

-----

**Indice dei contenuti**

<!-- MarkdownTOC autolink="true" bracket="round" autoanchor="false" lowercase="only_ascii" uri_encoding="true" levels="1,2,3" -->

- [Introduzione](#introduzione)
    - [Alan IF](#alan-if)
        - [Programmazione prosaica](#programmazione-prosaica)
        - [Libertà autoriale](#libert%C3%A0-autoriale)
    - [Alan Italian vs Libreria Standard Italiana](#alan-italian-vs-libreria-standard-italiana)
    - [Alan Standard Library](#alan-standard-library)
- [Stato del progetto](#stato-del-progetto)
- [Sostieni il progetto](#sostieni-il-progetto)
- [Ultime novità](#ultime-novit%C3%A0)
    - [2019/02/28: Adozione di Highlight per la documentazione](#20190228-adozione-di-highlight-per-la-documentazione)
    - [2019/02/27: Passaggio a Alan 3.0beta6 build 1878](#20190227-passaggio-a-alan-30beta6-build-1878)
    - [2019/02/21: Passaggio a Alan 3.0beta6 build 1870](#20190221-passaggio-a-alan-30beta6-build-1870)
    - [2019/02/19: Riscritto il codice per il vestiario](#20190219-riscritto-il-codice-per-il-vestiario)
    - [2019/02/07: Inizio lavori per il "Ricettario di Alan"](#20190207-inizio-lavori-per-il-ricettario-di-alan)
    - [2019/02/05: Modulo per l'italiano indipendente e pronto all'uso](#20190205-modulo-per-litaliano-indipendente-e-pronto-alluso)
    - [2019/02/01: Passaggio a Alan 3.0beta6 build 1866](#20190201-passaggio-a-alan-30beta6-build-1866)
    - [2019/01/24: Passaggio a Alan 3.0beta6 build 1862](#20190124-passaggio-a-alan-30beta6-build-1862)
    - [2019/01/17: Passaggio a Alan 3.0beta6 build 1855](#20190117-passaggio-a-alan-30beta6-build-1855)
    - [2019/01/15: Grossa svolta con Alan 3.0beta6 build 1852](#20190115-grossa-svolta-con-alan-30beta6-build-1852)
        - [Supporto per le preposizioni apostrofate](#supporto-per-le-preposizioni-apostrofate)
        - [Baco della e-grave risolto!](#baco-della-e-grave-risolto)
    - [2018/11/13: Tutti i verbi tradotti](#20181113-tutti-i-verbi-tradotti)
- [Contenuti del progetto](#contenuti-del-progetto)
- [Impostare l'ambiente di lavoro](#impostare-lambiente-di-lavoro)
    - [Encoding dei file](#encoding-dei-file)
- [Ringraziamenti](#ringraziamenti)
- [Componenti terze parti](#componenti-terze-parti)
    - [Base16](#base16)
    - [Asciidoctor](#asciidoctor)
        - [Attributi italiani Asciidoctor](#attributi-italiani-asciidoctor)
    - [Estensione per Highlight](#estensione-per-highlight)
        - [ERB Templates](#erb-templates)
    - [Sass Boilerplate](#sass-boilerplate)
- [Link di approfondimento](#link-di-approfondimento)

<!-- /MarkdownTOC -->

-----

# Introduzione

Questo progetto si prefigge due obiettivi:

1. Tradurre in italiano la _[Alan Standard Library]_ v2.1 di [Anssi Räisänen]. 
2. Fornire un modulo indipendente per il supporto della lingua italiana in Alan IF 3.

Entrambi consentiranno la creazione di avventure testuali in lingua italiana utilizzando Alan 3, un linguaggio dedicato alla creazione di opere di interactive fiction.

Il primo dei due obiettivi è mirato a fornire ai novizi di Alan uno strumento di facile utilizzo per la creazione di avventure in italiano, grazie a una ricca infrastruttura di classi e verbi comunemente utilizzati nelle avventure.

Il secondo, è mirato agli utenti più esperti di Alan, che vogliono creare le proprie avventure partendo solo con il supporto per la lingua italiana, e costruirsi da zero tutti i verbi e l'infrastruttura dell'avventura.


## Alan IF

- https://www.alanif.se/
- https://bitbucket.org/alanif/alan

Alan (acronimo per __Adventure LANguage__) è un linguaggio dedicato alla creazione di [avventure testuali] (dette anche "AT", o "IF"/"Interactive Fiction"). Creato da Thomas Nilsson e Göran Forslund nel lontano 1985. Alan è oggi giunto alla sua terza incarnazione (Alan 3). Sebbene la versione 3 di Alan sia ancora in fase Beta, è sufficientemente stabile da poter essere usata per la creazione di avventure.

Gli ideatori di Alan hanno voluto creare uno strumento che fosse semplice da usare per chi non avesse esperienza di programmazione, e che non imponesse un modello standard su come debba essere strutturato il mondo delle avventure. Queste caretterstiche rendono Alan un ottimo candidato per la creazione di avventure in italiano.

### Programmazione prosaica

La sintassi di Alan è modellata attorno alle esigenze dello scrittore, più che del programmatore, quindi l'uso di parentesi, punteggiatura e altri simboli per l'organizzazione sintattica del codice è ridotto al minimo indispensabile, e non vi sono vincoli strutturali riguardo l'uso di indentazione e la ripartizione del codice nelle righe del sorgente. L'aspetto di un file sorgente di Alan è più simile alla prosa che non alla programmazione.

Questo è un esempio (completo) di codice Alan per creare un'avventura con una stanza, un attore ed un verbo, scritto in forma prosaica:

```alan
The classroom IsA location
Name 'English Literature Classroom'. Exit south to corridor.
Description "there are no studends in the classroom, except you."
End the classroom.

Syntax greet = greet (person) where person IsA actor else "You can only greet people."

The teacher IsA actor at classroom.
Name teacher. Name professor mr John Keating. Pronoun him.
Description "Mr Keating is busy at his desk, proofreading the students' homework."
Verb greet does
     "-- Good evening professor Keating.
    $nA whole minute goes by before your teacher finally raises
      his head from the papers and acknowledges your presence.
    $n-- How may I help you, Mr. Anderson?"
End verb. End the teacher.

The corridor IsA location
  EXit north to classroom.
End the corridor.

Start at classroom.
```

Questo è il medesimo codice, scritto in una forma che è più tipica della programmazione:

```alan
THE classroom IsA location
  NAME 'English Literature Classroom'.
  DESCRIPTION "There are no studends in the classroom, except you."
  EXIT south TO corridor.
END THE classroom.

SYNTAX greet = greet (person)
  WHERE person IsA actor
    ELSE "You can only greet people."

THE teacher IsA actor AT classroom.
  NAME teacher.
  NAME professor mr John Keating.
  PRONOUN him.
  DESCRIPTION
    "Mr Keating is busy at his desk, proofreading the students' homework."
  VERB greet
    DOES
      "-- Good evening professor Keating. $nA whole minute goes by before your
       teacher finally raises his head from the papers and acknowledges your
       presence. $n-- How may I help you, Mr. Anderson?"
  END VERB greet.
END THE teacher.

THE corridor IsA location
  EXIT north TO classroom.
END THE corridor.

START AT classroom.
```

In entrambi i casi, anche un lettore che non conoscesse Alan sarebbe in grado di intuire a cosa serve il codice, e come funzionerà l'avventura da esso creato. Il primo esempio risulterà più facile da comprendere a chi non ha esperienza di programmazione, mentre il secondo risulterà più familiare ai programmatori. Ogni autore è libero di adottare lo stile che preferisce nel proprie avventure.

Nei sorgenti della libreria di questo progetto, adotteremo uno stile di programmazione formale, più simile al secondo esempio, per ragioni di praticità di editing. Nei sorgenti degli esempi adotteremo invece uno stile meno formale, più prosaico.

Sebbene Alan adotti un approccio sintattico alla programmazione che ne semplifica l'uso, avvicinandola il più possibile all'uso naturale della lingua inglese, si tratta comunque di un vero e proprio linguaggio di programmazione formale — da non confodere, per esempio, con la [programmazione in lingua naturale] adottata da [Inform 7].

### Libertà autoriale

A differenza di altri sistemi per la creazione di AT, Alan non impone un modello standard del mondo dell'avventura, o un set di verbi predefiniti; la filosofia di Alan è di lasciare all'autore totale autonomia su come modellare l'avventura. Alan mette a disposizione un insieme di funzionalità e meccanismi che sono comuni a tutte le avventure, ma starà ai singoli autori decidere se e come usarle — ad esempio le funzionalità per il salvataggio e il caricamento di una partita sono controllate dall parole chiave `Save` e `Restore`, ma sarà l'autore a dover decidere _se e come_ rendere queste funzionalità accessibili al giocatore (p.es. implementando i comandi per il giocatore "salva" e "carica").


Per gli autori che sono abituati a lavorare con Inform, TADS, e altri sistemi di autoraggio dotati di una libreria standard che mette a disposizione i verbi, le classi ed i messaggi d'uso comune, l'approccio _tabula rasa_ di Alan potrà sembrare insolito. Lavorare con Alan è come dipingere partendo da una tela bianca: toccherà a noi scegliere i pennelli ed i colori da usare, e preparare la tela dandogli un fondo appropriato al tipo di colori che useremo. Gli altri sistemi, invece, sono come un kit di partenza dove il materiale di lavoro è già presente: abbiamo pennelli di varie misure, un set di tubetti con tutti i colori principali, e una tela già pretrattata e pronta all'uso.

A secondo del tipo di avventura che si vuole creare, i modelli preconfezionati potranno risultare troppo dettagliati o troppo scarni — molto dipende dal tipo di realismo che vogliamo conseguire. Magari non ci servono affatto le funzionalità per il vestiario o per i dialoghi, oppure ce ne servirebbe una versione più dettagliata e realistica. Alan ci consente di costruire un'infrastruttura _ad hoc_ per le nostre avventure, partendo da zero.

Alan mette a disposizione le seguenti classi di base con cui modellare il mondo delle AT:

![Le classi predefinite di Alan][classi predefinite]

Si tratta delle classi universali che sono alla base di qualsiasi avventura, e ciascuna di esse ha caratteristiche specifiche studiate appositamente per l'interactive fiction. L'autore potrà estenderle e specializzarle secondo necessità — creando supporti, animali, persone, indumenti, cibi e bevande, armi, e quant'altro. Queste classi sono come i colori primari dell'artista, e usandole saggiamente sarà possibile illustrare mondi fantastici ricchi e vivaci, creando le giuste tinte e sfumature con cui narrare la propria avventura.


## Alan Italian vs Libreria Standard Italiana

Per gli amanti della libertà assoluta nel plasmare il mondo delle avventure, __Alan__ "nudo e crudo" è lo strumento ideale, e __Alan Italian__ è un modulo a sé stante che aggiunge il supporto della grammatica italiana senza intaccare la _tabula rasa_ offerta da Alan.

Per chi invece preferisce scrivere le proprie avventure avvalendosi di un modello ricco e consolidato, la __Alan Standard Library__ è lo strumento ideale da cui partire, rendendo l'uso di Alan simile agli altri sistemi di autoraggio. Per gli autori italiani, questo progetto mette a disposizione
la __Libreria Standard Italiana__, che è la traduzione/adattamento della libreria inglese.

## Alan Standard Library

- https://www.alanif.se/information/library
- https://github.com/AnssiR66/AlanStdLib (nuovo repository ufficiale)
- https://bitbucket.org/alanif/alanlib

La _Alan Standard Library_, scritta da [Anssi Räisänen], definisce un ricco insieme di verbi, classi e funzionalità comunemente usati nelle avventure testuali, che gli autori possono usare come fondamenta su cui costruire le proprie avventure.

La libreria standard non è inclusa nella distribuzione ufficiale di Alan, e va scaricata separatamente.

Questo progetto è finalizzato alla traduzione italiana della _Alan Standard Library_.

# Stato del progetto

Non mi è ancora possibile fornire una data prevista per l'ultimazione della libreria, dato che vi sto lavoro a singhiozzo, a seconda del tempo libero a disposizione (sicuramente ci vorrà ancora parecchio tempo).

Ho comunque scelto di condividere su GitHub il progetto sin dall'inizio, nel caso qualcuno fosse interessanto a seguirne gli sviluppi e contribuirvi attivamente. I contributi e le opinoni di terzi sarebero di grande aiuto per accellerare il progetto; quindi sarei grato a chiunque volesse contribuire clonando il repository, sperimentando con la libreria, condvidendo opinioni, consigli e segnalandomi bachi ed errori. 

Potete seguire la discussione di questo progetto all'interno del [gruppo di discussione Alan-IF su Yahoo Groups][Alan Yahoo] (in inglese). Per qualsiasi domanda e suggerimento, potete [aprire uno issue qui].

# Sostieni il progetto

Da aprile 2018 ad oggi sono state investite centinaia di ore di lavoro (ed ettolitri di caffé) in questo progetto open source. Sostieni il progetto con una donazione.

[![Donazione PayPal][donazione img]][donazione link]

E non dimenticarti di dare una stella al progetto!

Grazie.


# Ultime novità

Alcune notizie flash sugli aggiornamenti importanti del progetto...

## 2019/02/28: Adozione di Highlight per la documentazione

<img src="https://assets.gitlab-static.net/uploads/-/system/project/avatar/6678916/highlight_256.png" align="left" width=192 alt="Highlight logo" title="Logo di Highlight, il coloratore di sintassi di André Simon">

D'ora in poi il progetto utilizzerà lo strumento [Highlight] di [André Simon] (anziché [highlight.js]) per la colorazione sintattica del codice nella documentazione del progetto.

Highlight è uno strumento molto più flessibile di [highlight.js], consente un maggior controllo sulle definizioni delle sintassi e — sebbene attualmente l'estensione di Highlight per Asciidoctor non supporti i callout o la colorazione di codice all'interno di tabelle (ci sto lavorando) — Highlight offre numerosi vantaggi:

<br clear="all" />

- I documenti HTML non dipenderanno più da JavaScript.
- Supporto per la numerazione delle righe del codice.
- Estensibilità delle funzionalità tramite plugin Lua.
- Numerosi formati di output supportati oltre allo HTML (XHTML, RTF, LaTeX, TeX, SVG, BBCode, e altri).
- [Oltre 210 linguaggi di programmazione e markup supportati].
- Circa 200 temi colorati nativi.


## 2019/02/27: Passaggio a Alan 3.0beta6 build 1878

La libreria adotta la nuova [developer snaphshot] di Alan [3.0beta6 build 1878].

## 2019/02/21: Passaggio a Alan 3.0beta6 build 1870

La libreria adotta la nuova [developer snaphshot] di Alan [3.0beta6 build 1870].


## 2019/02/19: Riscritto il codice per il vestiario

A partire dalla versione v0.15.0, la Libreria adotta un nuovo sistema per la gestione del vestiario. Adesso l'uso del vestiario da parte dell'utente finale sarà più semplice ed intuitivo.

Il codice riguardante gli indumenti è stato praticamente riscritto tutto, per ovviare a una serie di problematiche e bachi presenti nella libreria originale.

Queste modifiche consentiranno inoltre di introdurre a breve nuove funzionalità per il vestiario, che ne renderanno l'uso nelle avventure ancora più flessibile e personalizzabile.


## 2019/02/07: Inizio lavori per il "Ricettario di Alan"

- [`/ricettario/`](./ricettario)


alanlib_ita/CHANGELOG.md
# CHANGELOG

    Status: Alpha stage | Log Range: Alpha v0.12.1–.

Changelog of Italian translation of Alan Standard Library v2.1.

This is the current changelog from Alpha v0.12.1 upward — the CHANGELOG was split into two files due to its huge size.

For previuos changes, see:

- [`CHANGELOG_OLD.md`](./CHANGELOG_OLD.md)  (v0.0.1–0.12.1)


-----

**Indice**

<!-- MarkdownTOC autolink="true" bracket="round" autoanchor="false" lowercase="only_ascii" uri_encoding="true" levels="1,2,3" -->

- [Alan 3.0beta6 build 2022](#alan-30beta6-build-2022)
    - [2019/08/25](#20190825)
        - [Verbi](#verbi)
        - [Eventi](#eventi)
        - [Messaggi libreria](#messaggi-libreria)
    - [2019/08/24 \(2\)](#20190824-2)
        - [Verbi](#verbi-1)
        - [Messaggi libreria](#messaggi-libreria-1)
    - [2019/08/24 \(1\)](#20190824-1)
        - [Migliorie sintassi verbi](#migliorie-sintassi-verbi)
        - [Messaggi libreria](#messaggi-libreria-2)
        - [Risposte verbi](#risposte-verbi)
    - [2019/08/22](#20190822)
        - [Passa alla v0.22.0](#passa-alla-v0220)
        - [Passa ad Alan 3.0beta6 build 2022](#passa-ad-alan-30beta6-build-2022)
        - [Pulizia Sorgenti](#pulizia-sorgenti)
        - [Modulo classi](#modulo-classi)
- [Alan 3.0beta6 build 2015](#alan-30beta6-build-2015)
    - [2019/08/19](#20190819)
        - [Correggi e migliora verbo `ascolta`](#correggi-e-migliora-verbo-ascolta)
    - [2019/08/18](#20190818)
        - [Sinonimi "Dentro" e "Sopra"](#sinonimi-dentro-e-sopra)
        - [Problemi vari](#problemi-vari)
    - [2019/08/15](#20190815)
        - [Passa alla v0.21.0](#passa-alla-v0210)
        - [Passa ad Alan 3.0beta6 build 2015](#passa-ad-alan-30beta6-build-2015)
- [Alan 3.0beta6 build 1980](#alan-30beta6-build-1980)
    - [2019/07/28](#20190728)
        - [Liquidi](#liquidi)
        - [Contenitori Elencati](#contenitori-elencati)
        - [Messaggi Libreria](#messaggi-libreria-3)
    - [2019/06/26](#20190626)
        - [Repulisti classi attori](#repulisti-classi-attori)
        - [Verbo `prendi_da` sempre onnipotente](#verbo-prendi_da-sempre-onnipotente)
- [Alan 3.0beta6 build 1980](#alan-30beta6-build-1980-1)
    - [2019/06/16](#20190616)
        - [Nuovi messaggi dei verbi](#nuovi-messaggi-dei-verbi)
        - [Traduci messaggi dei verbi](#traduci-messaggi-dei-verbi)
        - [Elimina messaggi dei verbi](#elimina-messaggi-dei-verbi)
        - [Verbo `prendi_da`](#verbo-prendi_da)
    - [2019/06/15 \(3\)](#20190615-3)
        - [Elimina Verbi Modalità Lunga/Breve](#elimina-verbi-modalit%C3%A0-lungabreve)
    - [2019/06/15 \(2\)](#20190615-2)
        - [Sinonimi Direzioni](#sinonimi-direzioni)
        - [Doxterizza Modulo Luoghi](#doxterizza-modulo-luoghi)
    - [2019/06/15 \(1\)](#20190615-1)
        - [Passa alla v0.20.0](#passa-alla-v0200)
        - [Passa ad Alan 3.0beta6 build 1980](#passa-ad-alan-30beta6-build-1980)
- [Alan 3.0beta6 build 1880](#alan-30beta6-build-1880)
    - [2019/04/02](#20190402)
        - [Modulo Italian unificato](#modulo-italian-unificato)
    - [2019/03/23](#20190323)
        - [Passa alla v0.18.0](#passa-alla-v0180)
        - [Passa ad Alan 3.0beta6 build 1880](#passa-ad-alan-30beta6-build-1880)
        - [Riferimenti di versione](#riferimenti-di-versione)
        - [Modifiche Banner](#modifiche-banner)
- [Alan 3.0beta6 build 1878](#alan-30beta6-build-1878)
    - [2019/02/27](#20190227)
        - [Passa alla v0.17.0](#passa-alla-v0170)
        - [Passa ad Alan 3.0beta6 build 1878](#passa-ad-alan-30beta6-build-1878)
- [Alan 3.0beta6 build 1870](#alan-30beta6-build-1870)
    - [2019/02/21](#20190221)
        - [Passa alla v0.16.0](#passa-alla-v0160)
        - [Passa ad Alan 3.0beta6 build 1870](#passa-ad-alan-30beta6-build-1870)
        - [Aggiornamento vestiario](#aggiornamento-vestiario)
- [Alan 3.0beta6 build 1866](#alan-30beta6-build-1866)
    - [2019/02/19](#20190219)
        - [Passa alla v0.15.0](#passa-alla-v0150)
        - [Rinnovato il codice del vestiario](#rinnovato-il-codice-del-vestiario)
        - [Novità riguardanti il vestiario](#novit%C3%A0-riguardanti-il-vestiario)
        - [Nuovi messaggi dei verbi](#nuovi-messaggi-dei-verbi-1)
    - [2019/02/05](#20190205)
        - [Passa alla v0.14.0](#passa-alla-v0140)
        - [Separa modulo grammatica italiana](#separa-modulo-grammatica-italiana)
        - [Modulo classi](#modulo-classi-1)
        - [Modulo messaggi runtime](#modulo-messaggi-runtime)
    - [2019/02/02 \(2\)](#20190202-2)
        - [Passa alla v0.13.0](#passa-alla-v0130)
        - [Separa modulo classi vestiario](#separa-modulo-classi-vestiario)
        - [Vestiario](#vestiario)
    - [2019/02/02 \(1\)](#20190202-1)
        - [Passa ad Alan 3.0beta6 build 1866](#passa-ad-alan-30beta6-build-1866)

<!-- /MarkdownTOC -->

-----

# Alan 3.0beta6 build 2022

## 2019/08/25

- [`lib_messaggi_libreria.i`][lib_messaggi_libreria] &#x27f6; v0.22.3
- [`lib_luoghi.i`][lib_luoghi] &#x27f6; v0.22.1
- [`lib_verbi.i`][lib_verbi] &#x27f6; v0.22.3

### Verbi

#### Risposte azioni autodirette

Aggiungi messaggi _ad hoc_ nei seguenti verbi se indirizzati al protagonista:

|        verbo        |                   risposta                   |
|---------------------|----------------------------------------------|
| `guarda_in`         | "Ora come ora non è tempo di introspezione." |
| `guarda_attraverso` | `mia_AT:azione_insensata`                    |

### Eventi

Tradotti identificativi eventi:

|  evento EN  |   evento IT   |   in modulo    |
|-------------|---------------|----------------|
| `light_off` | `luci_spente` | `lib_luoghi.i` |

### Messaggi libreria

Eliminati attributi stringa originali inglesi che non erano più utilizzati da alcun verbo:

- `check_obj_not_in_worn2`


Rimpiazzati attributi stringa con testo in loco nel corpo del verbo:

|       attributo       |      verbo      |
|-----------------------|-----------------|
| `check_obj_not_hero5` | `libera`        |
| `check_obj_not_hero7` | `guarda_dietro` |

verbump.sh
#!/bin/bash

script="verbump.sh"
version="1.0.0"
revdate="2019/04/02"

echo -e "\e[94m\n******************************************************************************"
echo -e "*                                                                            *"
echo -e "*                         \e[93mAGGIORNA INFO DI VERSIONE\e[94m                          *"
echo -e "*                                                                            *"
echo -e "******************************************************************************"
echo -e "\e[90m\"$script\" v$version ($revdate) by Tristano Ajmone"
  echo -e "------------------------------------------------------------------------------\e[97m"
. ./ver.sh

echo -e "Tutti i riferimenti nel codice e nella documentazione verranno aggiornati ai"
echo -e "seguenti valori:\n"
echo -e "- Alan SDK: \e[93m$AlanSDK\e[97m"
echo -e "- Alan Ita: \e[93m$AlanITA ($AlanITA_RevDate)\n\e[97m"

# ==============================================================================
# 1. SETUP
# ==============================================================================
shopt -s globstar

# scomponi versione Alan Italian:

AlanITA_MAJOR=`echo $AlanITA | grep -o -P '^\d+'`
AlanITA_MINOR=`echo $AlanITA | grep -o -P '^\d+\.\K\d+'`
AlanITA_PATCH=`echo $AlanITA | grep -o -P '^(\d+\.){2}\K\d+'`

# -----------------------------------
# definisci stringhe per sostituzioni
# -----------------------------------

# Alan Ita (lunga): v0.17.0-Alpha
subAlanItaL="v$AlanITA-Alpha"

# Alan Ita (breve): Alan Italian 17 Alpha
subAlanItaS="Alan Italian $AlanITA_MINOR Alpha"

# Ala Ita Data (YYYY-MM-DD)
subData=$AlanITA_RevDate


# Alan Ver: 3.0beta6 build 1878
subAlanVer="$AlanSDK"

# Alan SDK: Alan 3.0beta6 build 1878
subAlanSDK="Alan $AlanSDK"

# -------------------------------------
# definisci alcune espressioni regolari
# -------------------------------------

# Le righe di commento con le info di versione hanno una di queste due forme:

: <<'ESEMPIO' [ Alan ]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
--| v0.17.0-Alpha, 2019-02-27: Alan 3.0beta6 build 1878
-- Alan Italian 17 Alpha | Alan 3.0beta6 build 1878
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ESEMPIO

# Alan Ita (lunga): v0.17.0-Alpha
rexAlanItaL="v[0-9]+\.[0-9]+\.[0-9]+-Alpha"

# Alan Ita (breve): Alan Italian 17 Alpha
rexAlanItaS="Alan Italian [0-9]+ Alpha"

# Ala Ita Data (YYYY-MM-DD)
rexData="[0-9]{4}-[0-9]{2}-[0-9]{2}"

# Alan Ver: 3.0beta6 build 1878
rexAlanVer="3\.0beta[0-9]+( build [0-9]{4})?"

# Alan SDK: Alan 3.0beta6 build 1878
rexAlanSDK="Alan $rexAlanVer"

# -------------------------
# definisci alcune funzioni
# -------------------------

function Heading1 {
  echo -e "\e[93m******************************************************************************"
  printf  "\e[94m%*s\n" $(((${#1}+78)/2)) "$1"
  echo -e "\e[93m******************************************************************************\e[97m"
}

function Heading2 {
  echo -e "\e[94m=============================================================================="
  echo -e "\e[95m$1"
  echo -e "\e[94m==============================================================================\e[97m" 
}

function ProcessaFile {
  # $1 : files pattern
  # $2 : find RegEx (PCRE)
  # $2 : substitution string
  # $4 : lines range (optional)

  files=$1
  files2fix="$files2fix $files"

  echo -e "\n\e[94mSED File Option: \e[93m$1"
  echo -e   "\e[94mSED Lines Range: \e[93m$4"
  echo -e   "\e[94mSED RegExp Find: \e[93m$2"
  echo -e   "\e[94mSED RegExp Repl: \e[93m$3"

  for f in $files ; do
  
    # Anteprima su STDOUT
    Separatore
    echo -e "$f:\e[92m"
    sed -s -n -E \
      -e "$4 s/$2/$3/p" \
      $f

    # Modifica i file:
    sed -i -E \
      -e "$4 s/$2/$3/" \
      $f
  done
  outEnd
}

function Separatore {
  # riga grigio scuro
  echo -e "\e[90m------------------------------------------------------------------------------"
}
function outBeg {
  # riga grigio scuro + imposta colore verde
  echo -e "\e[90m------------------------------------------------------------------------------\e[92m"
}
function outEnd {
  # riga grigio scuro + imposta colore bianco
  echo -e "\e[90m------------------------------------------------------------------------------\e[97m"
}
# ==============================================================================
# 2. AGGIORNA INFO DI VERSIONE
# ==============================================================================

Heading1 "AGGIORNA INFO DI VERSIONE"

# ------------------------------------------------------------------------------
# README.adoc
# ------------------------------------------------------------------------------
# Modifica attributi AsciiDoc di versione in "README.adoc":

: <<'ESEMPIO' [ AsciiDoc ]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
= Alan Italian
Tristano Ajmone <tajmone@gmail.com>
2019-03-01
// tag::AlanRevInfo[]
:AlanItaV: 17
:AlanSDK: 3.0beta6 build 1878
:revnumber: 0.{AlanItaV}-Alpha
// end::AlanRevInfo[]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ESEMPIO

# NOTA: Questi attributi sono posizionati su numeri di riga fissi (5, 6).
#       Si noti anche che sono racchiusi in una regione taggata di modo da poter
#       essere inclusi selettivamente dagli altri documenti del progetto durante
#       la conversione HTML, fungendo così da definizione centralizzata delle
#       informazioni di versione Alan per l'intera documentazione del progetto.
       
Heading2 "Attributi AsciiDoc in \e[93m./README.adoc"
echo -e "Vanno aggiornati due attributi: '\e[93mAlanItaV\e[97m' e '\
\e[93mAlanSDK\e[97m'."

ProcessaFile \
  "./README.adoc" \
  "^(:AlanItaV: )[0-9]+" \
  "\1$AlanITA_MINOR" \
  "5"

ProcessaFile \
  "./README.adoc" \
  "^(:AlanSDK: )$rexAlanVer" \
  "\1$subAlanVer" \
  "6"

# ------------------------------------------------------------------------------
# alanlib_ita/lib_definizioni.i
# ------------------------------------------------------------------------------
# Modifica attributi Alan di versione nei sorgenti della Libreria:

: <<'ESEMPIO' [ Alan ]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  HAS         AlanV  "3.0beta6 build 1878".
  HAS      AlanItaV  "17 Alpha".
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ESEMPIO

Heading2 "Attributi Alan nella Libreria"
echo -e "Vanno aggiornati due attributi: '\e[93mAlanV\e[97m' e '\
\e[93mAlanItaV\e[97m'."

ProcessaFile \
  "alanlib_ita/lib_definizioni.i" \
  "^( *HAS +AlanV +\")$rexAlanVer" \
  "\1$subAlanVer"

ProcessaFile \
  "alanlib_ita/lib_definizioni.i" \
  "^( *HAS +AlanItaV +\")[0-9]+( Alpha)" \
  "\1$AlanITA_MINOR\2"


# ------------------------------------------------------------------------------
# moduli della libreria
# ------------------------------------------------------------------------------
# L'intestazione dei moduli della libreria adotta questa convenzione:

: <<'ESEMPIO' [ Alan ]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
--= Alan StdLib Italian: Verbi
--| Tristano Ajmone <tajmone@gmail.com>
--~-----------------------------------------------------------------------------
--~ "lib_verbi.i"
--| v0.17.0-Alpha, 2019-02-27: Alan 3.0beta6 build 1878
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ESEMPIO

# la riga che ci interessa è sempre la quinta riga.

Heading2 "Moduli libreria"
echo -e "Va aggiornata la riga di commento con le info di versione. "

ProcessaFile \
  "./alanlib_ita/*.i" \
  "^--\| $rexAlanItaL, $rexData: $rexAlanSDK" \
  "--| $subAlanItaL, $subData: $subAlanSDK" \
  "5"

# ------------------------------------------------------------------------------
# avventure
# ------------------------------------------------------------------------------
# I sorgenti delle avventure e dei moduli aggiuntivi in genere contengono le
# informazioni di versione nella prima riga commentata:

: <<'ESEMPIO' [ Alan ]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-- Alan Italian 17 Alpha | Alan 3.0beta6 build 1878
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ESEMPIO

# la riga che ci interessa è sempre e solo la prima riga.

Heading2 "Sorgenti avventure"
echo -e "Va aggiornata la prima riga commentato dello header con le info di versione. "

ProcessaFile \
  "./test/**/*.alan ./test/**/*.i\
   ./demo/**/*.alan\
   ./ricettario/**/*.alan" \
  "^-- $rexAlanItaS \| $rexAlanSDK" \
  "-- $subAlanItaS | $subAlanSDK" \
  "1"

# ------------------------------------------------------------------------------
# avventure doxterizzate
# ------------------------------------------------------------------------------
# Alcune avventure demo o della test suite hanno attributi AsciiDoc di versione.

: <<'ESEMPIO' [ Alan + Doxter ]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-->rev_info
--| :revdate:     2019-02-20
--| :revnumber:   Alpha
--| :AlanSDK:     Alan 3.0beta6 build 1878
--<<
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ESEMPIO

# NOTA: La posizione di questi attributi non è fissa dato che sorgenti diversi
#       avranno header diversi. Ma l'attributo sarà sempre entro le prime 20
#       righe, quindi un range '1,20' è sicuro da usare.
       
Heading2 "Attributi AsciiDoc/Doxter in sorgenti avventure"
echo -e "Va aggiornato l'attributo \e[93mAlanSDK\e[97m' nei commenti Doxter. "

ProcessaFile \
  "./demo/**/*.alan ./test/vestiario/ega.alan" \
  "^(--\| :AlanSDK: +Alan )$rexAlanVer" \
  "\1$subAlanVer" \
  "1,20"

# ==============================================================================
# 3. POST-PROCESSING
# ==============================================================================

# ------------------------------------------------------------------------------
# correggi EOL (solo windows)
# ------------------------------------------------------------------------------

# Perché in Bash for Windows sed modifica tutte le EOL in LF.

if [[ $(uname -s) == MINGW* ]];then
  Heading1 "CORREGGI EOL"
  echo -e "Correggi sequenze EOL da LF a CRLF:"
  echo -e "\e[90m------------------------------------------------------------------------------\e[32m"
  unix2dos $files2fix
  echo -e "\e[90m------------------------------------------------------------------------------\e[32m"
fi

exit
